File: Go/tree/tree.go
package tree

import (
	"bytes"
	"fmt"
	"math"

	. "game_of_stones/common"
)

type Game[move Equatable[move]] interface {
	Turn() Turn
	TopMoves(*[]MoveValue[move])
	PlayMove(move)
	UndoMove(move)
	Decision() Decision
	BoardValue() int16
}

type Tree[move Equatable[move]] struct {
	game     Game[move]
	nodes    []node
	moves    []move
	topMoves []MoveValue[move]
	history  []move
}

type node struct {
	firstChild int32
	lastChild  int32
	nSims      int32
	value      int16
	decision   Decision
}

const explorationFactor = 20

func NewTree[move Equatable[move]](
	game Game[move],
) *Tree[move] {
	var m move
	return &Tree[move]{
		game:  game,
		nodes: []node{{}},
		moves: []move{m},
	}
}

func (tree *Tree[m]) Expand() (decision Decision, forcedMove bool) {
	root := &tree.nodes[0]
	if root.decision == NoDecision {
		tree.expand(0)
		tree.validate()
	}

	undecided := 0
	for i := root.firstChild; i < root.lastChild; i++ {
		child := tree.nodes[i]
		if child.decision == NoDecision {
			if child.nSims > 1 {
				undecided++
			} else {
				return root.decision, false
			}
		}
	}
	return root.decision, undecided == 1
}

func (tree *Tree[move]) CommitMove(toPlay move) {
	tree.game.PlayMove(toPlay)
	tree.history = append(tree.history, toPlay)

	tree.nodes = tree.nodes[:0]
	tree.nodes = append(tree.nodes, node{
		value:    tree.game.BoardValue(),
		decision: tree.game.Decision(),
	})
	tree.moves = tree.moves[:0]
	tree.moves = append(tree.moves, toPlay)
}

func (tree *Tree[move]) Reset() {
	lastMove := tree.history[len(tree.history)-1]
	tree.history = tree.history[:len(tree.history)-1]
	tree.game.UndoMove(lastMove)

	tree.nodes = tree.nodes[:0]
	tree.nodes = append(tree.nodes, node{
		value:    tree.game.BoardValue(),
		decision: tree.game.Decision(),
	})
	var m move
	tree.moves = tree.moves[:0]
	tree.moves = append(tree.moves, m)
}

func (tree *Tree[move]) BestMove() move {
	root := tree.nodes[0]

	if tree.game.Turn() == First && root.decision == SecondWin ||
		tree.game.Turn() == Second && root.decision == FirstWin {

		for idx := root.firstChild; idx < root.lastChild; idx++ {
			node := tree.nodes[idx]
			if node.lastChild-node.firstChild == 1 {
				return tree.moves[node.firstChild]
			}
		}
	}

	bestChildIdx := root.firstChild

	if tree.game.Turn() == First {
		for idx := root.firstChild; idx < root.lastChild; idx++ {
			node := tree.nodes[idx]
			best := tree.nodes[bestChildIdx]
			if best.decision == FirstWin {
				if node.decision == FirstWin && best.nSims < node.nSims {
					bestChildIdx = idx
				}
			} else if best.decision == SecondWin {
				if node.decision != SecondWin || best.nSims < node.nSims {
					bestChildIdx = idx
				}
			} else if node.decision == FirstWin || best.value < node.value {
				bestChildIdx = idx
			}
		}
	} else {
		for idx := root.firstChild; idx < root.lastChild; idx++ {
			node := tree.nodes[idx]
			best := tree.nodes[bestChildIdx]
			if best.decision == SecondWin {
				if node.decision == SecondWin && best.nSims < node.nSims {
					bestChildIdx = idx
				}
			} else if best.decision == FirstWin {
				if node.decision != FirstWin || best.nSims < node.nSims {
					bestChildIdx = idx
				}
			} else if node.decision == SecondWin || best.value > node.value {
				bestChildIdx = idx
			}
		}
	}

	return tree.moves[bestChildIdx]
}

func (tree *Tree[m]) AvailableMoves() string {
	buf := &bytes.Buffer{}
	root := tree.nodes[0]
	fmt.Fprintf(buf, "%s: d: %v v: %v n: %d\n", tree.moves[0].String(), root.decision, root.value, root.nSims)
	for i := root.firstChild; i < root.lastChild; i++ {
		child := tree.nodes[i]
		fmt.Fprintf(buf, "  [%2d] %s: d: %v v: %v n: %d\n", i, tree.moves[i].String(), child.decision, child.value, child.nSims)
	}
	return buf.String()
}

func (tree *Tree[m]) Value() int16 {
	return tree.nodes[0].value
}

func (tree *Tree[m]) expand(parentIdx int32) {
	parent := &tree.nodes[parentIdx]

	if parent.firstChild == 0 {
		tree.game.TopMoves(&tree.topMoves)
		if len(tree.topMoves) == 0 {
			panic("Function top_moves(game, ...) returns empty result.")
		}

		parent.firstChild = int32(len(tree.nodes))
		parent.lastChild = int32(len(tree.nodes) + len(tree.topMoves))
		for _, childMoveValue := range tree.topMoves {
			tree.nodes = append(tree.nodes, node{
				nSims:    1,
				value:    childMoveValue.Value,
				decision: childMoveValue.Decision,
			})
			tree.moves = append(tree.moves, childMoveValue.Move)
		}
	} else {
		var coeff float64 = 1
		if tree.game.Turn() == Second {
			coeff = -1
		}
		selectedChildIdx := int32(-1)
		logParentSims := math.Log2(float64(parent.nSims))
		maxV := math.Inf(-1)
		for idx := parent.firstChild; idx < parent.lastChild; idx++ {
			child := tree.nodes[idx]
			if child.decision != NoDecision {
				continue
			}
			v := coeff*float64(child.value) + explorationFactor*math.Sqrt(logParentSims/float64(child.nSims))
			if v > maxV {
				maxV = v
				selectedChildIdx = idx
			}
		}

		tree.game.PlayMove(tree.moves[selectedChildIdx])
		tree.expand(selectedChildIdx)
		tree.game.UndoMove(tree.moves[selectedChildIdx])
	}

	parent = &tree.nodes[parentIdx]
	parent.nSims = int32(0)
	parent.value = tree.nodes[parent.firstChild].value
	decision := NoDecision
	if tree.game.Turn() == First {
		b_win := false
		w_win := true
		all_draws := true
		for i := parent.firstChild; i < parent.lastChild; i++ {
			child := tree.nodes[i]
			parent.nSims += child.nSims
			parent.value = max(parent.value, child.value)
			b_win = b_win || child.decision == FirstWin
			w_win = w_win && child.decision == SecondWin
			all_draws = all_draws && (child.decision == Draw || child.decision == SecondWin)
		}
		if b_win {
			decision = FirstWin
		} else if w_win {
			decision = SecondWin
		} else if all_draws {
			decision = Draw
		}
	} else {
		w_win := false
		b_win := true
		all_draws := true
		for i := parent.firstChild; i < parent.lastChild; i++ {
			child := tree.nodes[i]
			parent.nSims += child.nSims
			parent.value = min(parent.value, child.value)
			w_win = w_win || child.decision == SecondWin
			b_win = b_win && child.decision == FirstWin
			all_draws = all_draws && (child.decision == Draw || child.decision == FirstWin)
		}
		if w_win {
			decision = SecondWin
		} else if b_win {
			decision = FirstWin
		} else if all_draws {
			decision = Draw
		}
	}

	parent.decision = decision
}

func (tree *Tree[move]) String() string {
	buf := &bytes.Buffer{}
	tree.string(buf, 0, 0)
	return buf.String()
}

func (tree *Tree[move]) string(buf *bytes.Buffer, idx int32, depth int) {
	buf.WriteRune('\n')
	for range depth {
		buf.WriteString("|   ")
	}

	fmt.Fprint(buf, tree.moves[idx])
	fmt.Fprintf(buf, " [%d] ", idx)
	node := tree.nodes[idx]
	node.string(buf)
	for childIdx := node.firstChild; childIdx < node.lastChild; childIdx++ {
		tree.string(buf, childIdx, depth+1)
	}
}

func (node *node) String() string {
	buf := &bytes.Buffer{}
	node.string(buf)
	return buf.String()
}

func (node *node) string(buf *bytes.Buffer) {
	fmt.Fprintf(buf, "v: %d s: %d d: %v", node.value, node.nSims, node.decision)
}

File: Swift/Sources/Tree/Tree.swift
import Foundation

protocol GameProtocol {
    associatedtype Move: MoveProtocol
    
    mutating func topMoves(_: [Move])
    mutating func playMove(_ move: Move)
    mutating func undoMove()
    func decision() -> Decision
}

public enum Decision: CustomStringConvertible { 
    case NoDecision, FirstWin, SecondWin, Draw

    public var description: String {
        return switch self {
            case .NoDecision: "no-decision"
            case .FirstWin: "first-win"
            case .SecondWin: "second-win"
            case .Draw: "draw"
        }
    }
}

protocol MoveProtocol: Comparable {
    init()

    var value: Float { get set }

    var isWin: Bool { get set }
    var isLoss: Bool { get set }
    var isDraw: Bool { get set }
    var isDecisive: Bool { get }
}

struct Tree<Game: GameProtocol> {
    var game: Game
    let c: Float
    var root: Node<Game> = Node<Game>(Game.Move())
    var topMoves: [Game.Move] = []
    var bestMove: Game.Move { root.bestMove }

    init(game: Game, c: Float) {
        self.game = game
        self.c = c
    }

    mutating func expand() -> Bool {
        if root.move.isDecisive {
            return true
        } else {
            root.expand(&game, c, &topMoves)
        }
        
        if root.move.isDecisive {
            return true
        }

        var undecided = 0
        for child in root.children {
            if !child.move.isDecisive {
                undecided += 1
            }
        }
        return undecided == 1
    }

    mutating func reset() {
        root = Node<Game>(Game.Move())
    }
}

struct Node<Game: GameProtocol> {
    var move: Game.Move
    var children: [Self] = []
    var nSims: Int32 = 1

    init(_ move: Game.Move) {
        self.move = move
    }

    mutating func expand(_ game: inout Game, _ c: Float, _ topMoves: inout [Game.Move]) {
        if children.isEmpty {
            game.topMoves(topMoves)
            assert(!topMoves.isEmpty, "Function top<oves(...) returns empty result.")

            children.reserveCapacity(topMoves.count)
            for move in topMoves {
                children.append(Node(move))
            }
        } else {
            var selectedChildIdx = 0
            let nSims = nSims
            let logParentSims = log2f(Float(nSims))
            var maxV = -Float.infinity
            for childIdx in children.indices {
                if children[childIdx].move.isDecisive {
                    continue
                }
                let v = children[childIdx].move.value + c * sqrt(logParentSims / Float(children[childIdx].nSims))
                if v > maxV {
                    maxV = v
                    selectedChildIdx = childIdx
                }
            }
            game.playMove(children[selectedChildIdx].move)
            children[selectedChildIdx].expand(&game, c, &topMoves)
            game.undoMove()
        }

        nSims = 0
        move.value = Float.infinity
        var hasDraw = false
        var allDraws = true
        for child in children {
            if child.move.isWin {
                move.isLoss = true
                return
            } else if child.move.isDraw {
                hasDraw = true
                continue
            }
            allDraws = false
            if child.move.isLoss {
                continue
            }
            nSims += child.nSims
            if move.value >= -child.move.value {
                move.value = -child.move.value
            }
        }
        if allDraws {
            move.isDraw = true
        } else if hasDraw && move.value > 0 {
            move.value = 0
        }
    }

    var bestMove: Game.Move {
        assert(!children.isEmpty, "Function node.bestMove() is called with no children.")
        var bestChildIdx = 0
        for childIdx in children.indices {
            if children[bestChildIdx].move.value < children[childIdx].move.value {
                bestChildIdx = childIdx
            } else if children[bestChildIdx].move.isLoss && children[bestChildIdx].nSims < children[childIdx].nSims {
                bestChildIdx = childIdx
            }
        }
        return children[bestChildIdx].move
    }

}


